// next-ws@2.2.5/dist/server/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/server/index.ts
var server_exports = {};
__export(server_exports, {
  getAdapter: () => getAdapter,
  getHttpServer: () => getHttpServer,
  getWebSocketServer: () => getWebSocketServer,
  setAdapter: () => setAdapter,
  setHttpServer: () => setHttpServer,
  setWebSocketServer: () => setWebSocketServer,
  setupWebSocketServer: () => setupWebSocketServer
});
module.exports = __toCommonJS(server_exports);

// src/server/persistent.ts
function useGlobal(key) {
  return [
    function get() {
      return Reflect.get(globalThis, key);
    },
    function set(value) {
      return Reflect.set(globalThis, key, value);
    },
    function use(getter) {
      const existing = Reflect.get(globalThis, key);
      if (existing) return existing;
      Reflect.set(globalThis, key, getter());
      return Reflect.get(globalThis, key);
    }
  ];
}
var [getHttpServer, setHttpServer, useHttpServer] = (
  //
  useGlobal(
    /* @__PURE__ */ Symbol.for("next-ws.http-server")
    //
  )
);
var [getWebSocketServer, setWebSocketServer, useWebSocketServer] = (
  //
  useGlobal(
    /* @__PURE__ */ Symbol.for("next-ws.websocket-server")
    //
  )
);
var [getRequestStorage, setRequestStorage, useRequestStorage] = (
  //
  useGlobal(
    /* @__PURE__ */ Symbol.for("next-ws.request-store")
    //
  )
);
var [getAdapter, setAdapter, useAdapter] = (
  //
  useGlobal(
    /* @__PURE__ */ Symbol.for("next-ws.adapter")
    //
  )
);

// src/server/setup.ts
var import_node_async_hooks = require("async_hooks");
var logger2 = __toESM(require("next/dist/build/output/log.js"));
var ws = __toESM(require("ws"));

// src/server/helpers/match.ts
function compileRoutePattern(routePattern) {
  const escapedPattern = routePattern.replace(/[-/\\^$*+?.()|[\]{}]/g, "\\$&");
  const paramRegex = escapedPattern.replace(/\\\[\\\[(?:\\\.){3}([a-z0-9_]+)\\\]\\\]/gi, "?(?<r_o_$1>.+)?").replace(/\\\[(?:\\\.){3}([a-z0-9_]+)\\\]/gi, "(?<r_$1>.+)").replace(/\\\[([a-z0-9_]+)\\\]/gi, "(?<$1>[^/]+)");
  return new RegExp(`^${paramRegex}$`);
}
function getRouteParams(routePattern, requestPathname) {
  const routeRegex = compileRoutePattern(routePattern);
  const match = requestPathname.replace(/\/+$/, "").match(routeRegex);
  if (!match) return null;
  if (!match.groups) return {};
  const params = {};
  for (let [k, v] of Object.entries(match.groups)) {
    if (k.startsWith("r_")) {
      const optional = k.startsWith("r_o_");
      k = k.slice(optional ? 4 : 2);
      v = v?.split("/");
    }
    if (v) Reflect.set(params, k, v);
  }
  return params;
}
function findMatchingRoute(nextServer, requestPathname) {
  const basePath = nextServer.serverOptions?.conf.basePath || "";
  const appPathRoutes = {
    // @ts-expect-error - appPathRoutes is protected
    ...nextServer.appPathRoutes,
    // @ts-expect-error - getAppPathRoutes is protected
    ...nextServer.getAppPathRoutes()
  };
  const sortedRoutes = Object.entries(appPathRoutes).filter(([routePath, file]) => routePath && file?.length).sort(([a], [b]) => scoreRoute(b) - scoreRoute(a));
  for (const [routePath, [filePath]] of sortedRoutes) {
    if (!filePath) continue;
    const realPath = `${basePath}${routePath}`;
    const routeParams = getRouteParams(realPath, requestPathname);
    if (routeParams) {
      return { filename: filePath, params: routeParams };
    }
  }
  return void 0;
}
function scoreRoute(routePath) {
  const parts = routePath.split("/").filter(Boolean);
  let score = 0;
  if (parts.length === 0) score = Infinity;
  for (const part of parts) {
    if (part.startsWith("[[...")) score += -1;
    else if (part.startsWith("[...")) score += 0;
    else if (part.startsWith("[")) score += 2;
    else score += 3;
  }
  return score + parts.length;
}

// src/server/helpers/module.ts
var logger = __toESM(require("next/dist/build/output/log.js"));
async function importRouteModule(nextServer, filePathname) {
  try {
    if ("hotReloader" in nextServer) {
      await nextServer.hotReloader?.ensurePage({
        page: filePathname,
        clientOnly: false
      });
    } else if ("ensurePage" in nextServer) {
      await nextServer.ensurePage({ page: filePathname, clientOnly: false });
    } else {
      logger.warnOnce(
        "[next-ws] unable to ensure page, you may need to open the route in your browser first so Next.js compiles it"
      );
    }
  } catch {
  }
  try {
    const buildPagePath = nextServer.getPagePath(filePathname);
    return (await require(buildPagePath)).routeModule;
  } catch (cause) {
    console.error(cause);
    return void 0;
  }
}

// src/server/helpers/request.ts
var import_server = require("next/server");
function toNextRequest(message) {
  const controller = new AbortController();
  const headers = new Headers(message.headers);
  const protocol = "encrypted" in message.socket ? "https" : "http";
  const url = `${protocol}://${headers.get("host")}${message.url}`;
  message.once("aborted", () => controller.abort());
  return new import_server.NextRequest(url, {
    method: message.method,
    headers,
    body: message.method === "GET" || message.method === "HEAD" ? void 0 : message,
    signal: controller.signal,
    referrer: headers.get("referer") || void 0
  });
}

// src/server/helpers/store.ts
var import_cookies = require("next/dist/compiled/@edge-runtime/cookies");
var ReadonlyHeaders = class extends Headers {
  append() {
    throw new Error("Headers are read-only");
  }
  set() {
    throw new Error("Headers are read-only");
  }
  delete() {
    throw new Error("Headers are read-only");
  }
};
var ReadonlyRequestsCookies = class extends import_cookies.RequestCookies {
  set() {
    throw new Error("Cookies are read-only");
  }
  delete() {
    throw new Error("Cookies are read-only");
  }
};
function createRequestStore(request) {
  return {
    headers: new ReadonlyHeaders(request.headers),
    cookies: new ReadonlyRequestsCookies(request.headers)
  };
}

// src/server/setup.ts
var WebSocketServer2 = ws.WebSocketServer || ws.Server;
function setupWebSocketServer(nextServer, options) {
  const httpServer = (
    //
    // @ts-expect-error - serverOptions is protected
    useHttpServer(() => nextServer.serverOptions?.httpServer)
  );
  if (!httpServer)
    return logger2.error("[next-ws] was not able to find the HTTP server");
  const wsServer = (
    //
    useWebSocketServer(() => new WebSocketServer2({ noServer: true }))
  );
  const requestStorage = (
    //
    useRequestStorage(() => new import_node_async_hooks.AsyncLocalStorage())
  );
  const adapter = options?.adapter;
  if (adapter) {
    logger2.ready("[next-ws] adapter configured for multi-instance support");
  } else {
    logger2.ready("[next-ws] has started the WebSocket server");
  }
  const kAttached = /* @__PURE__ */ Symbol.for("next-ws.http-server.attached");
  if (Reflect.has(httpServer, kAttached)) return;
  Reflect.set(httpServer, kAttached, true);
  const routeClients = /* @__PURE__ */ new Map();
  httpServer.on("upgrade", async (message, socket, head) => {
    const request = toNextRequest(message);
    const pathname = request.nextUrl.pathname;
    if (pathname.includes("/_next")) return;
    const route = findMatchingRoute(nextServer, pathname);
    if (!route) {
      logger2.error(`[next-ws] could not find route for page ${pathname}`);
      return socket.end();
    }
    const module2 = await importRouteModule(nextServer, route.filename);
    if (!module2) {
      logger2.error(`[next-ws] could not import module for page ${pathname}`);
      return socket.end();
    }
    const handleUpgrade = module2.userland.UPGRADE;
    const handleSocket = module2.userland.SOCKET;
    if ((!handleUpgrade || typeof handleUpgrade !== "function") && (!handleSocket || typeof handleSocket !== "function")) {
      logger2.error(`[next-ws] route '${pathname}' does not export a handler`);
      return socket.end();
    }
    if (handleSocket)
      logger2.warnOnce(
        "DeprecationWarning: [next-ws] SOCKET is deprecated, prefer UPGRADE instead, see https://github.com/k0d13/next-ws#-usage"
      );
    wsServer.handleUpgrade(message, socket, head, async (client) => {
      wsServer.emit("connection", client, message);
      if (adapter) {
        if (!routeClients.has(pathname)) {
          routeClients.set(pathname, /* @__PURE__ */ new Set());
          await adapter.onMessage(pathname, (message2) => {
            const clients = routeClients.get(pathname);
            if (!clients) return;
            for (const localClient of clients) {
              if (localClient.readyState === ws.WebSocket.OPEN) {
                localClient.send(
                  typeof message2 === "string" ? message2 : JSON.stringify(message2)
                );
              }
            }
          });
        }
        routeClients.get(pathname)?.add(client);
      }
      try {
        const context = { params: route.params };
        if (handleUpgrade) {
          await requestStorage.run(
            createRequestStore(request),
            //
            () => handleUpgrade(client, wsServer, request, context)
          );
        } else if (handleSocket) {
          const handleClose = (
            //
            await handleSocket(client, message, wsServer, context)
          );
          if (typeof handleClose === "function")
            client.once("close", () => handleClose());
        }
        if (adapter) {
          client.on("message", (data) => {
            adapter.broadcast(pathname, data).catch((err) => {
              logger2.error("[next-ws] adapter broadcast failed:", err);
            });
          });
        }
      } catch (cause) {
        logger2.error(
          `[next-ws] error in socket handler for '${pathname}'`,
          cause
        );
        try {
          client.close(1011, "Internal Server Error");
        } catch {
        }
      }
      client.once("close", () => {
        if (adapter) {
          routeClients.get(pathname)?.delete(client);
        }
      });
    });
    return;
  });
  if (adapter) {
    httpServer.once("close", async () => {
      await adapter.close().catch((err) => {
        logger2.error("[next-ws] adapter cleanup failed:", err);
      });
    });
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getAdapter,
  getHttpServer,
  getWebSocketServer,
  setAdapter,
  setHttpServer,
  setWebSocketServer,
  setupWebSocketServer
});
