// next-ws@2.2.5/dist/cli.cjs
#!/usr/bin/env node
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// src/commands/helpers/define.ts
var import_minimist = __toESM(require("minimist"));

// package.json
var version = "2.2.5";

// src/commands/helpers/define.ts
function defineCommandGroup(definition) {
  return {
    ...definition,
    parse(parents, argv) {
      const parsed = (0, import_minimist.default)(argv);
      for (const child of this.children)
        if (parsed._[0] === child.name)
          return void child.parse(parents.concat(this), argv.slice(1));
      if (parsed.help)
        return void console.log(buildCommandGroupHelp(parents, this));
      if (parsed.v || parsed.version) return void console.log(version);
      return void console.log(buildCommandGroupHelp(parents, this));
    }
  };
}
function buildCommandGroupHelp(parents, group) {
  return `Usage: ${[...parents, group].map((p) => p.name).join(" ")} [command] [options]

${group.description}

Commands:
  ${group.children.map((c) => `${c.name} | ${c.description}`).join("\n  ")}

Options:
  --help | Show this help message and exit.
  --version | Show the version number and exit.
`;
}
function defineCommand(definition) {
  return {
    ...definition,
    parse(parents, argv) {
      const parsed = (0, import_minimist.default)(argv);
      if (parsed.help) return void console.log(buildCommandHelp(parents, this));
      return this.action(parsed);
    }
  };
}
function buildCommandHelp(parents, command) {
  return `Usage: ${[...parents, command].map((p) => p.name).join(" ")} [options]

${command.description}

Options:
  --help | Show this help message and exit.
  ${command.options.map((o) => `--${o.name} | ${o.description}`).join("\n  ")}
`;
}

// src/patches/helpers/next.ts
var import_promises = require("fs/promises");
var import_node_path = require("path");
function resolveNextWsDirectory() {
  const id = (
    //
    require.resolve("next-ws/package.json", { paths: [process.cwd()] })
  );
  return (0, import_node_path.dirname)(id);
}
function resolveNextDirectory() {
  const id = (
    //
    require.resolve("next/package.json", { paths: [process.cwd()] })
  );
  return (0, import_node_path.dirname)(id);
}
async function getInstalledNextVersion() {
  const id = (0, import_node_path.join)(resolveNextDirectory(), "package.json");
  const pkg = await (0, import_promises.readFile)(id, "utf8").then(JSON.parse);
  return String(pkg.version.split("-")[0]);
}
async function readTrace() {
  const id = (0, import_node_path.join)(resolveNextDirectory(), ".next-ws-trace.json");
  return (0, import_promises.readFile)(id, "utf-8").then(JSON.parse).catch(() => null);
}
async function writeTrace(trace) {
  const id = (0, import_node_path.join)(resolveNextDirectory(), ".next-ws-trace.json");
  await (0, import_promises.writeFile)(id, JSON.stringify(trace, null, 2));
}

// src/patches/patch-1.ts
var import_node_path3 = require("path");
var import_jscodeshift = __toESM(require("jscodeshift"));

// src/patches/helpers/define.ts
var import_promises2 = require("fs/promises");
var import_node_path2 = require("path");

// src/commands/helpers/console.ts
var import_node_readline = __toESM(require("readline"));
var import_node_util = require("util");

// node_modules/.pnpm/chalk@5.6.2/node_modules/chalk/source/vendor/ansi-styles/index.js
var ANSI_BACKGROUND_OFFSET = 10;
var wrapAnsi16 = (offset = 0) => (code) => `\x1B[${code + offset}m`;
var wrapAnsi256 = (offset = 0) => (code) => `\x1B[${38 + offset};5;${code}m`;
var wrapAnsi16m = (offset = 0) => (red, green, blue) => `\x1B[${38 + offset};2;${red};${green};${blue}m`;
var styles = {
  modifier: {
    reset: [0, 0],
    // 21 isn't widely supported and 22 does the same thing
    bold: [1, 22],
    dim: [2, 22],
    italic: [3, 23],
    underline: [4, 24],
    overline: [53, 55],
    inverse: [7, 27],
    hidden: [8, 28],
    strikethrough: [9, 29]
  },
  color: {
    black: [30, 39],
    red: [31, 39],
    green: [32, 39],
    yellow: [33, 39],
    blue: [34, 39],
    magenta: [35, 39],
    cyan: [36, 39],
    white: [37, 39],
    // Bright color
    blackBright: [90, 39],
    gray: [90, 39],
    // Alias of `blackBright`
    grey: [90, 39],
    // Alias of `blackBright`
    redBright: [91, 39],
    greenBright: [92, 39],
    yellowBright: [93, 39],
    blueBright: [94, 39],
    magentaBright: [95, 39],
    cyanBright: [96, 39],
    whiteBright: [97, 39]
  },
  bgColor: {
    bgBlack: [40, 49],
    bgRed: [41, 49],
    bgGreen: [42, 49],
    bgYellow: [43, 49],
    bgBlue: [44, 49],
    bgMagenta: [45, 49],
    bgCyan: [46, 49],
    bgWhite: [47, 49],
    // Bright color
    bgBlackBright: [100, 49],
    bgGray: [100, 49],
    // Alias of `bgBlackBright`
    bgGrey: [100, 49],
    // Alias of `bgBlackBright`
    bgRedBright: [101, 49],
    bgGreenBright: [102, 49],
    bgYellowBright: [103, 49],
    bgBlueBright: [104, 49],
    bgMagentaBright: [105, 49],
    bgCyanBright: [106, 49],
    bgWhiteBright: [107, 49]
  }
};
var modifierNames = Object.keys(styles.modifier);
var foregroundColorNames = Object.keys(styles.color);
var backgroundColorNames = Object.keys(styles.bgColor);
var colorNames = [...foregroundColorNames, ...backgroundColorNames];
function assembleStyles() {
  const codes = /* @__PURE__ */ new Map();
  for (const [groupName, group] of Object.entries(styles)) {
    for (const [styleName, style] of Object.entries(group)) {
      styles[styleName] = {
        open: `\x1B[${style[0]}m`,
        close: `\x1B[${style[1]}m`
      };
      group[styleName] = styles[styleName];
      codes.set(style[0], style[1]);
    }
    Object.defineProperty(styles, groupName, {
      value: group,
      enumerable: false
    });
  }
  Object.defineProperty(styles, "codes", {
    value: codes,
    enumerable: false
  });
  styles.color.close = "\x1B[39m";
  styles.bgColor.close = "\x1B[49m";
  styles.color.ansi = wrapAnsi16();
  styles.color.ansi256 = wrapAnsi256();
  styles.color.ansi16m = wrapAnsi16m();
  styles.bgColor.ansi = wrapAnsi16(ANSI_BACKGROUND_OFFSET);
  styles.bgColor.ansi256 = wrapAnsi256(ANSI_BACKGROUND_OFFSET);
  styles.bgColor.ansi16m = wrapAnsi16m(ANSI_BACKGROUND_OFFSET);
  Object.defineProperties(styles, {
    rgbToAnsi256: {
      value(red, green, blue) {
        if (red === green && green === blue) {
          if (red < 8) {
            return 16;
          }
          if (red > 248) {
            return 231;
          }
          return Math.round((red - 8) / 247 * 24) + 232;
        }
        return 16 + 36 * Math.round(red / 255 * 5) + 6 * Math.round(green / 255 * 5) + Math.round(blue / 255 * 5);
      },
      enumerable: false
    },
    hexToRgb: {
      value(hex) {
        const matches = /[a-f\d]{6}|[a-f\d]{3}/i.exec(hex.toString(16));
        if (!matches) {
          return [0, 0, 0];
        }
        let [colorString] = matches;
        if (colorString.length === 3) {
          colorString = [...colorString].map((character) => character + character).join("");
        }
        const integer = Number.parseInt(colorString, 16);
        return [
          /* eslint-disable no-bitwise */
          integer >> 16 & 255,
          integer >> 8 & 255,
          integer & 255
          /* eslint-enable no-bitwise */
        ];
      },
      enumerable: false
    },
    hexToAnsi256: {
      value: (hex) => styles.rgbToAnsi256(...styles.hexToRgb(hex)),
      enumerable: false
    },
    ansi256ToAnsi: {
      value(code) {
        if (code < 8) {
          return 30 + code;
        }
        if (code < 16) {
          return 90 + (code - 8);
        }
        let red;
        let green;
        let blue;
        if (code >= 232) {
          red = ((code - 232) * 10 + 8) / 255;
          green = red;
          blue = red;
        } else {
          code -= 16;
          const remainder = code % 36;
          red = Math.floor(code / 36) / 5;
          green = Math.floor(remainder / 6) / 5;
          blue = remainder % 6 / 5;
        }
        const value = Math.max(red, green, blue) * 2;
        if (value === 0) {
          return 30;
        }
        let result = 30 + (Math.round(blue) << 2 | Math.round(green) << 1 | Math.round(red));
        if (value === 2) {
          result += 60;
        }
        return result;
      },
      enumerable: false
    },
    rgbToAnsi: {
      value: (red, green, blue) => styles.ansi256ToAnsi(styles.rgbToAnsi256(red, green, blue)),
      enumerable: false
    },
    hexToAnsi: {
      value: (hex) => styles.ansi256ToAnsi(styles.hexToAnsi256(hex)),
      enumerable: false
    }
  });
  return styles;
}
var ansiStyles = assembleStyles();
var ansi_styles_default = ansiStyles;

// node_modules/.pnpm/chalk@5.6.2/node_modules/chalk/source/vendor/supports-color/index.js
var import_node_process = __toESM(require("process"), 1);
var import_node_os = __toESM(require("os"), 1);
var import_node_tty = __toESM(require("tty"), 1);
function hasFlag(flag, argv = globalThis.Deno ? globalThis.Deno.args : import_node_process.default.argv) {
  const prefix = flag.startsWith("-") ? "" : flag.length === 1 ? "-" : "--";
  const position = argv.indexOf(prefix + flag);
  const terminatorPosition = argv.indexOf("--");
  return position !== -1 && (terminatorPosition === -1 || position < terminatorPosition);
}
var { env } = import_node_process.default;
var flagForceColor;
if (hasFlag("no-color") || hasFlag("no-colors") || hasFlag("color=false") || hasFlag("color=never")) {
  flagForceColor = 0;
} else if (hasFlag("color") || hasFlag("colors") || hasFlag("color=true") || hasFlag("color=always")) {
  flagForceColor = 1;
}
function envForceColor() {
  if ("FORCE_COLOR" in env) {
    if (env.FORCE_COLOR === "true") {
      return 1;
    }
    if (env.FORCE_COLOR === "false") {
      return 0;
    }
    return env.FORCE_COLOR.length === 0 ? 1 : Math.min(Number.parseInt(env.FORCE_COLOR, 10), 3);
  }
}
function translateLevel(level) {
  if (level === 0) {
    return false;
  }
  return {
    level,
    hasBasic: true,
    has256: level >= 2,
    has16m: level >= 3
  };
}
function _supportsColor(haveStream, { streamIsTTY, sniffFlags = true } = {}) {
  const noFlagForceColor = envForceColor();
  if (noFlagForceColor !== void 0) {
    flagForceColor = noFlagForceColor;
  }
  const forceColor = sniffFlags ? flagForceColor : noFlagForceColor;
  if (forceColor === 0) {
    return 0;
  }
  if (sniffFlags) {
    if (hasFlag("color=16m") || hasFlag("color=full") || hasFlag("color=truecolor")) {
      return 3;
    }
    if (hasFlag("color=256")) {
      return 2;
    }
  }
  if ("TF_BUILD" in env && "AGENT_NAME" in env) {
    return 1;
  }
  if (haveStream && !streamIsTTY && forceColor === void 0) {
    return 0;
  }
  const min = forceColor || 0;
  if (env.TERM === "dumb") {
    return min;
  }
  if (import_node_process.default.platform === "win32") {
    const osRelease = import_node_os.default.release().split(".");
    if (Number(osRelease[0]) >= 10 && Number(osRelease[2]) >= 10586) {
      return Number(osRelease[2]) >= 14931 ? 3 : 2;
    }
    return 1;
  }
  if ("CI" in env) {
    if (["GITHUB_ACTIONS", "GITEA_ACTIONS", "CIRCLECI"].some((key) => key in env)) {
      return 3;
    }
    if (["TRAVIS", "APPVEYOR", "GITLAB_CI", "BUILDKITE", "DRONE"].some((sign) => sign in env) || env.CI_NAME === "codeship") {
      return 1;
    }
    return min;
  }
  if ("TEAMCITY_VERSION" in env) {
    return /^(9\.(0*[1-9]\d*)\.|\d{2,}\.)/.test(env.TEAMCITY_VERSION) ? 1 : 0;
  }
  if (env.COLORTERM === "truecolor") {
    return 3;
  }
  if (env.TERM === "xterm-kitty") {
    return 3;
  }
  if (env.TERM === "xterm-ghostty") {
    return 3;
  }
  if (env.TERM === "wezterm") {
    return 3;
  }
  if ("TERM_PROGRAM" in env) {
    const version2 = Number.parseInt((env.TERM_PROGRAM_VERSION || "").split(".")[0], 10);
    switch (env.TERM_PROGRAM) {
      case "iTerm.app": {
        return version2 >= 3 ? 3 : 2;
      }
      case "Apple_Terminal": {
        return 2;
      }
    }
  }
  if (/-256(color)?$/i.test(env.TERM)) {
    return 2;
  }
  if (/^screen|^xterm|^vt100|^vt220|^rxvt|color|ansi|cygwin|linux/i.test(env.TERM)) {
    return 1;
  }
  if ("COLORTERM" in env) {
    return 1;
  }
  return min;
}
function createSupportsColor(stream, options = {}) {
  const level = _supportsColor(stream, {
    streamIsTTY: stream && stream.isTTY,
    ...options
  });
  return translateLevel(level);
}
var supportsColor = {
  stdout: createSupportsColor({ isTTY: import_node_tty.default.isatty(1) }),
  stderr: createSupportsColor({ isTTY: import_node_tty.default.isatty(2) })
};
var supports_color_default = supportsColor;

// node_modules/.pnpm/chalk@5.6.2/node_modules/chalk/source/utilities.js
function stringReplaceAll(string, substring, replacer) {
  let index = string.indexOf(substring);
  if (index === -1) {
    return string;
  }
  const substringLength = substring.length;
  let endIndex = 0;
  let returnValue = "";
  do {
    returnValue += string.slice(endIndex, index) + substring + replacer;
    endIndex = index + substringLength;
    index = string.indexOf(substring, endIndex);
  } while (index !== -1);
  returnValue += string.slice(endIndex);
  return returnValue;
}
function stringEncaseCRLFWithFirstIndex(string, prefix, postfix, index) {
  let endIndex = 0;
  let returnValue = "";
  do {
    const gotCR = string[index - 1] === "\r";
    returnValue += string.slice(endIndex, gotCR ? index - 1 : index) + prefix + (gotCR ? "\r\n" : "\n") + postfix;
    endIndex = index + 1;
    index = string.indexOf("\n", endIndex);
  } while (index !== -1);
  returnValue += string.slice(endIndex);
  return returnValue;
}

// node_modules/.pnpm/chalk@5.6.2/node_modules/chalk/source/index.js
var { stdout: stdoutColor, stderr: stderrColor } = supports_color_default;
var GENERATOR = /* @__PURE__ */ Symbol("GENERATOR");
var STYLER = /* @__PURE__ */ Symbol("STYLER");
var IS_EMPTY = /* @__PURE__ */ Symbol("IS_EMPTY");
var levelMapping = [
  "ansi",
  "ansi",
  "ansi256",
  "ansi16m"
];
var styles2 = /* @__PURE__ */ Object.create(null);
var applyOptions = (object, options = {}) => {
  if (options.level && !(Number.isInteger(options.level) && options.level >= 0 && options.level <= 3)) {
    throw new Error("The `level` option should be an integer from 0 to 3");
  }
  const colorLevel = stdoutColor ? stdoutColor.level : 0;
  object.level = options.level === void 0 ? colorLevel : options.level;
};
var chalkFactory = (options) => {
  const chalk2 = (...strings) => strings.join(" ");
  applyOptions(chalk2, options);
  Object.setPrototypeOf(chalk2, createChalk.prototype);
  return chalk2;
};
function createChalk(options) {
  return chalkFactory(options);
}
Object.setPrototypeOf(createChalk.prototype, Function.prototype);
for (const [styleName, style] of Object.entries(ansi_styles_default)) {
  styles2[styleName] = {
    get() {
      const builder = createBuilder(this, createStyler(style.open, style.close, this[STYLER]), this[IS_EMPTY]);
      Object.defineProperty(this, styleName, { value: builder });
      return builder;
    }
  };
}
styles2.visible = {
  get() {
    const builder = createBuilder(this, this[STYLER], true);
    Object.defineProperty(this, "visible", { value: builder });
    return builder;
  }
};
var getModelAnsi = (model, level, type, ...arguments_) => {
  if (model === "rgb") {
    if (level === "ansi16m") {
      return ansi_styles_default[type].ansi16m(...arguments_);
    }
    if (level === "ansi256") {
      return ansi_styles_default[type].ansi256(ansi_styles_default.rgbToAnsi256(...arguments_));
    }
    return ansi_styles_default[type].ansi(ansi_styles_default.rgbToAnsi(...arguments_));
  }
  if (model === "hex") {
    return getModelAnsi("rgb", level, type, ...ansi_styles_default.hexToRgb(...arguments_));
  }
  return ansi_styles_default[type][model](...arguments_);
};
var usedModels = ["rgb", "hex", "ansi256"];
for (const model of usedModels) {
  styles2[model] = {
    get() {
      const { level } = this;
      return function(...arguments_) {
        const styler = createStyler(getModelAnsi(model, levelMapping[level], "color", ...arguments_), ansi_styles_default.color.close, this[STYLER]);
        return createBuilder(this, styler, this[IS_EMPTY]);
      };
    }
  };
  const bgModel = "bg" + model[0].toUpperCase() + model.slice(1);
  styles2[bgModel] = {
    get() {
      const { level } = this;
      return function(...arguments_) {
        const styler = createStyler(getModelAnsi(model, levelMapping[level], "bgColor", ...arguments_), ansi_styles_default.bgColor.close, this[STYLER]);
        return createBuilder(this, styler, this[IS_EMPTY]);
      };
    }
  };
}
var proto = Object.defineProperties(() => {
}, {
  ...styles2,
  level: {
    enumerable: true,
    get() {
      return this[GENERATOR].level;
    },
    set(level) {
      this[GENERATOR].level = level;
    }
  }
});
var createStyler = (open, close, parent) => {
  let openAll;
  let closeAll;
  if (parent === void 0) {
    openAll = open;
    closeAll = close;
  } else {
    openAll = parent.openAll + open;
    closeAll = close + parent.closeAll;
  }
  return {
    open,
    close,
    openAll,
    closeAll,
    parent
  };
};
var createBuilder = (self, _styler, _isEmpty) => {
  const builder = (...arguments_) => applyStyle(builder, arguments_.length === 1 ? "" + arguments_[0] : arguments_.join(" "));
  Object.setPrototypeOf(builder, proto);
  builder[GENERATOR] = self;
  builder[STYLER] = _styler;
  builder[IS_EMPTY] = _isEmpty;
  return builder;
};
var applyStyle = (self, string) => {
  if (self.level <= 0 || !string) {
    return self[IS_EMPTY] ? "" : string;
  }
  let styler = self[STYLER];
  if (styler === void 0) {
    return string;
  }
  const { openAll, closeAll } = styler;
  if (string.includes("\x1B")) {
    while (styler !== void 0) {
      string = stringReplaceAll(string, styler.close, styler.open);
      styler = styler.parent;
    }
  }
  const lfIndex = string.indexOf("\n");
  if (lfIndex !== -1) {
    string = stringEncaseCRLFWithFirstIndex(string, closeAll, openAll, lfIndex);
  }
  return openAll + string + closeAll;
};
Object.defineProperties(createChalk.prototype, styles2);
var chalk = createChalk();
var chalkStderr = createChalk({ level: stderrColor ? stderrColor.level : 0 });
var source_default = chalk;

// src/commands/helpers/console.ts
function log(...message) {
  console.log("[next-ws]", ...message);
}
function info(...message) {
  console.log(source_default.blue("[next-ws]"), ...message);
}
function warn(...message) {
  console.log(source_default.yellow("[next-ws]"), ...message);
}
function error(...message) {
  console.log(source_default.red("[next-ws]"), ...message);
}
var debug = (0, import_node_util.debuglog)("next-ws");
function success(...message) {
  console.log(source_default.green("[next-ws]", "\u2714"), ...message);
}
function failure(...message) {
  console.log(source_default.red("[next-ws]", "\u2716"), ...message);
}
async function confirm(...message) {
  const rl = import_node_readline.default.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  return new Promise((resolve2) => {
    const question = source_default.yellow("[next-ws]", ...message);
    const options = source_default.cyan("[y/N]");
    rl.question(`${question} ${options}`, (answer) => {
      const normalisedAnswer = answer.trim().toLowerCase();
      if (normalisedAnswer === "y") resolve2(true);
      else resolve2(false);
      rl.close();
    });
  });
}
async function task(promise, ...message) {
  process.stdout.write("\x1B[?25l");
  const spinnerChars = ["\u280B", "\u2819", "\u2839", "\u2838", "\u283C", "\u2834", "\u2826", "\u2827"];
  let spinnerIndex = 0;
  const spinnerInterval = setInterval(() => {
    import_node_readline.default.cursorTo(process.stdout, 0);
    const spinnerChar = spinnerChars[spinnerIndex++ % spinnerChars.length] ?? " ";
    process.stdout.write(source_default.cyan("[next-ws]", spinnerChar, ...message));
  }, 100);
  return promise.then((value) => {
    clearInterval(spinnerInterval);
    import_node_readline.default.cursorTo(process.stdout, 0);
    import_node_readline.default.clearLine(process.stdout, 0);
    success(...message);
    return value;
  }).catch((err) => {
    clearInterval(spinnerInterval);
    import_node_readline.default.cursorTo(process.stdout, 0);
    import_node_readline.default.clearLine(process.stdout, 0);
    failure(...message);
    throw err;
  }).finally(() => {
    process.stdout.write("\x1B[?25h");
  });
}

// src/patches/helpers/define.ts
function definePatch(definition) {
  return {
    ...definition,
    async execute() {
      for (const step of this.steps)
        await task(step.execute(), step.title);
    }
  };
}
function definePatchStep(definition) {
  return {
    ...definition,
    get path() {
      return resolvePath(definition.path);
    },
    async execute() {
      debug(`Applying '${this.title}' to '${this.path}'`);
      const code = await (0, import_promises2.readFile)(this.path, "utf8").then((code2) => this.transform(code2));
      await (0, import_promises2.writeFile)(this.path, code);
    }
  };
}
function resolvePath(path) {
  switch (path.split(":")[0]) {
    case "next": {
      const nextDirectory = resolveNextDirectory();
      const realPath = path.slice(5);
      return (0, import_node_path2.resolve)(nextDirectory, realPath);
    }
    default: {
      return path;
    }
  }
}

// src/patches/patch-1.ts
var CommentLine = import_jscodeshift.default.Comment;
var patchNextNodeServer = definePatchStep({
  title: "Add WebSocket server setup script to NextNodeServer constructor",
  path: "next:dist/server/next-server.js",
  async transform(code) {
    const marker = "@patch attach-websocket-server";
    const snippet = (0, import_jscodeshift.default)(`
      // ${marker}
      let nextWs;
      try { nextWs ??= require('next-ws/server') } catch {}
      try { nextWs ??= require(require.resolve('next-ws/server', { paths: [process.cwd()] }) )} catch {}
      try { nextWs ??= require('${resolveNextWsDirectory().replaceAll(import_node_path3.sep, "/").replaceAll("'", "\\'")}/dist/server/index.cjs') } catch {}
      const adapter = nextWs?.getAdapter?.();
      nextWs?.setupWebSocketServer(this, adapter ? { adapter } : undefined);
    `);
    const block = import_jscodeshift.default.blockStatement(snippet.nodes()[0].program.body);
    return (0, import_jscodeshift.default)(code).find(import_jscodeshift.default.ClassDeclaration, { id: { name: "NextNodeServer" } }).find(import_jscodeshift.default.MethodDefinition, { kind: "constructor" }).forEach(({ node }) => {
      const body = node.value.body.body;
      const existing = (0, import_jscodeshift.default)(body).find(CommentLine, { value: ` ${marker}` }).paths()[0];
      const idx = body.indexOf(existing?.parent.node);
      if (existing && idx > -1) body[idx] = block;
      else body.push(block);
    }).toSource();
  }
});
var patchRouterServer = definePatchStep({
  title: "Prevent Next.js from immediately closing WebSocket connections",
  path: "next:dist/server/lib/router-server.js",
  async transform(code) {
    return (0, import_jscodeshift.default)(code).find(import_jscodeshift.default.ReturnStatement).find(import_jscodeshift.default.CallExpression, {
      callee: {
        type: "MemberExpression",
        object: { type: "Identifier", name: "socket" },
        property: { type: "Identifier", name: "end" }
      }
    }).replaceWith((path) => {
      const expr = import_jscodeshift.default.unaryExpression("void", import_jscodeshift.default.literal(0));
      expr.comments = [import_jscodeshift.default.commentLine(` ${(0, import_jscodeshift.default)(path.node).toSource()}`)];
      return expr;
    }).toSource();
  }
});
var patchNextTypesPlugin = definePatchStep({
  title: "Add UPGRADE as allowed export from route modules",
  path: "next:dist/build/webpack/plugins/next-types-plugin/index.js",
  async transform(code) {
    return (0, import_jscodeshift.default)(code).find(import_jscodeshift.default.MemberExpression, {
      property: { name: "HTTP_METHODS" }
    }).at(0).replaceWith((path) => {
      if (path.parent.value.type === "SpreadElement") return path.node;
      return import_jscodeshift.default.arrayExpression([
        import_jscodeshift.default.spreadElement(path.node),
        import_jscodeshift.default.literal("UPGRADE"),
        import_jscodeshift.default.literal("SOCKET")
      ]);
    }).toSource();
  }
});
var patchHeaders = definePatchStep({
  title: "Add WebSocket contextual headers resolution to request headers",
  path: "next:dist/client/components/headers.js",
  async transform(code) {
    const marker = "@patch headers";
    const snippet = (0, import_jscodeshift.default)(`
      // ${marker}
      const kRequestStorage = Symbol.for('next-ws.request-store');
      const requestStorage = Reflect.get(globalThis, kRequestStorage);
      const contextualHeaders = requestStorage?.getStore()?.headers;
      if (contextualHeaders) return contextualHeaders;
    `);
    const block = import_jscodeshift.default.blockStatement(snippet.nodes()[0].program.body);
    return (0, import_jscodeshift.default)(code).find(import_jscodeshift.default.FunctionDeclaration, { id: { name: "headers" } }).forEach(({ node }) => {
      const body = node.body.body;
      const existing = (0, import_jscodeshift.default)(body).find(CommentLine, { value: ` ${marker}` }).paths()[0];
      const idx = body.indexOf(existing?.parent.node);
      if (existing && idx > -1) body[idx] = block;
      else body.unshift(block);
    }).toSource();
  }
});
var patchCookies = definePatchStep({
  title: "Add WebSocket contextual cookies resolution to request cookies",
  path: "next:dist/client/components/headers.js",
  async transform(code) {
    const marker = "@patch cookies";
    const snippet = (0, import_jscodeshift.default)(`
      // ${marker}
      const kRequestStorage = Symbol.for('next-ws.request-store');
      const requestStorage = Reflect.get(globalThis, kRequestStorage);
      const contextualCookies = requestStorage?.getStore()?.cookies;
      if (contextualCookies) return contextualCookies;
    `);
    const block = import_jscodeshift.default.blockStatement(snippet.nodes()[0].program.body);
    return (0, import_jscodeshift.default)(code).find(import_jscodeshift.default.FunctionDeclaration, { id: { name: "cookies" } }).forEach(({ node }) => {
      const body = node.body.body;
      const existing = (0, import_jscodeshift.default)(body).find(CommentLine, { value: ` ${marker}` }).paths()[0];
      const idx = body.indexOf(existing?.parent.node);
      if (existing && idx > -1) body[idx] = block;
      else body.unshift(block);
    }).toSource();
  }
});
var patch_1_default = definePatch({
  name: "patch-1",
  versions: ">=13.5.1 <=14.2.32",
  steps: [
    patchNextNodeServer,
    patchNextTypesPlugin,
    patchRouterServer,
    patchHeaders,
    patchCookies
  ]
});

// src/patches/patch-2.ts
var patchHeaders2 = definePatchStep({
  ...patchHeaders,
  // CHANGE(next@15): headers function was moved
  path: "next:dist/server/request/headers.js"
});
var patchCookies2 = definePatchStep({
  ...patchCookies,
  // CHANGE(next@15): cookies function was moved
  path: "next:dist/server/request/cookies.js"
});
var patch_2_default = definePatch({
  name: "patch-2",
  versions: ">=15.0.0 <=16.2.4",
  steps: [
    patchNextNodeServer,
    patchRouterServer,
    patchNextTypesPlugin,
    patchHeaders2,
    patchCookies2
  ]
});

// src/patches/index.ts
var patches_default = [patch_1_default, patch_2_default];

// src/commands/helpers/semver.ts
var import_range = __toESM(require("semver/classes/range.js"));
var import_semver = __toESM(require("semver/classes/semver.js"));
var import_gt = __toESM(require("semver/functions/gt.js"));
var import_satisfies = __toESM(require("semver/functions/satisfies"));
var import_gtr = __toESM(require("semver/ranges/gtr"));
var import_ltr = __toESM(require("semver/ranges/ltr"));
var import_min_version = __toESM(require("semver/ranges/min-version"));
function maxVersion(range, loose) {
  range = new import_range.default(range, loose);
  let maximumVersion = null;
  for (const comparators of range.set) {
    for (const {
      operator,
      semver: { version: version_ }
    } of comparators) {
      if (operator === ">" || operator === ">=") continue;
      const version2 = new import_semver.default(version_);
      if (operator === "<") {
        version2.patch--;
        version2.raw = version2.format();
      }
      if (!maximumVersion || (0, import_gt.default)(version2, maximumVersion))
        maximumVersion = version2;
    }
  }
  return maximumVersion;
}

// src/commands/patch.ts
var patch_default = defineCommand({
  name: "patch",
  description: "Patch the local Next.js installation to support WebSockets",
  options: [
    {
      name: "yes",
      description: "Skip confirmation prompt for unsupported versions",
      alias: "y"
    }
  ],
  async action(options) {
    const supported = patches_default.map((p) => p.versions).join(" || ");
    const minimum = import_min_version.default(supported)?.version ?? supported;
    const maximum = maxVersion(supported)?.version ?? supported;
    const current = await getInstalledNextVersion();
    if (import_ltr.default(current, minimum)) {
      error(
        `Next.js v${current} is not supported, a minimum of v${minimum} is required`
      );
      process.exit(1);
    }
    let patch = patches_default.find((p) => import_satisfies.default(current, p.versions));
    if (import_gtr.default(current, maximum)) {
      warn(
        `Next.js v${current} is not officially supported, a maximum of v${maximum} is recommended.`
      );
      const confirm2 = options.yes || await confirm("Are you sure you want to proceed?");
      if (confirm2) {
        patch = patches_default[patches_default.length - 1];
        info("Proceeding with the latest patch");
        log(
          "If you encounter any issues please report them at https://github.com/k0d13/next-ws/issues"
        );
      } else {
        error("Aborted");
        process.exit(1);
      }
    }
    if (!patch) {
      error(
        `Next.js v${current} is not supported, please upgrade to a version within the range '${supported}'`
      );
      process.exit(1);
    }
    info(`Patching Next.js v${current} with '${patch.versions}'`);
    await patch.execute();
    info("Saving patch trace file...");
    await writeTrace({ patch: patch.versions, version: current });
    info("All done!");
  }
});

// src/commands/verify.ts
var verify_default = defineCommand({
  name: "verify",
  description: "Verify that the local Next.js installation has been patched",
  options: [
    {
      name: "ensure",
      description: "If not patched, then run the patch command",
      alias: "e"
    }
  ],
  async action(options) {
    const trace = await readTrace();
    if (!trace) {
      if (options.ensure) {
        warn("Next.js has not been patched, running the patch command");
        return patch_default.action({ yes: true });
      } else {
        error(
          "Next.js has not been patched, you'll need to run the patch command"
        );
        process.exit(1);
      }
    }
    const current = await getInstalledNextVersion();
    if (current !== trace.version) {
      error(
        "Next.js has been patched with a different version, you'll need to run the patch command"
      );
      process.exit(1);
    }
    info("Next.js has been patched!");
  }
});

// src/commands/index.ts
var commands_default = defineCommandGroup({
  name: "next-ws",
  description: "Patch the local Next.js installation to support WebSockets",
  children: [patch_default, verify_default]
});

// src/cli.ts
commands_default.parse([], process.argv.slice(2));
